### PR TITLE
Add general constraints capabilities to Opt Module

### LINKED ISSUE #24930: Add general constraint support for the optimization module
## Reason

Certain optimizations need general constraint supports that bounds can't satisfy. For example, volume constraint in topology optimization. 
## Design

Setup TAO callbacks and routines needed to apply constraints.
## Impact

Will allow users to perform more complex optimizations.

### PR DESCRIPTION
**Pull request by maxnezdyur**
closes #24930
Add general constraint capabilities from TAO

### CODE REVIEW

**Comment by GiudGiud** [comment]
@zachmprince

**Review by lynnmunday — CHANGES_REQUESTED** [review]
I didn't finish the review but his is what I have so far.  I still need to review the changes to optimize solve and the new reporters.

File: modules/optimization/doc/content/source/optimizationreporters/GeneralOptimization.md [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+!syntax description /OptimizationReporter/GeneralOptimization
+
+## Overview
+
+
+## Optimization Parameters
+
+`GeneralOptimization` is also responsible for creating parameter vector(s) for
+optimization, setting the initial condition for the optimization, and setting
+parameter bounds and constraints. Although the [Optimize.md] executioner holds a single vector
```
[the remark this answers is not in the document]
Comment (maxnezdyur) [reply modules/optimization/doc/content/source/optimizationreporters/GeneralOptimization.md:10]: There are a few additons here that are different from OptimizaitonReporter. For
example, we have `num_values` and `num_values_name`. Here we can set those
values with reporter, but that wouldn't make sense for OptRep where you know the
number of values ahead of time always because its related to the measurement data.

File: modules/optimization/src/optimizationreporters/OptimizationReporterBase.C [hunk 2]
```diff
@@ -138,3 +165,47 @@ OptimizationReporterBase::fillParamsVector(std::string type, Real default_value)
 
   return flattened_data;
 }
+
+void
+OptimizationReporterBase::computeEqualityConstraints(
```
Comment (lynnmunday) [inline modules/optimization/src/optimizationreporters/OptimizationReporterBase.C:170]: I think these should all be named "get" instead of "compute"

File: modules/optimization/src/reporters/ConstraintGradCollector.C [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+                                                                   REPORTER_MODE_REPLICATED));
+    }
+    params_ind++;
+  }
+}
+
+void
+ConstraintGradCollector::execute()
+{
+  const size_t num_params = _parameters.size();
```
Comment (lynnmunday) [inline modules/optimization/src/reporters/ConstraintGradCollector.C:64]: I was told to use std::size_t

File: modules/optimization/test/src/optimizationreporters/QuadraticMinimizeConstrained.C [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+  unsigned int i = 0;
+  for (const auto & param : _parameters)
+  {
+    Real equality_constraint = 0;
+    for (const auto & val : *param)
+    {
+      equality_constraint += val;
+    }
+    eqs_constraints.set(i++, equality_constraint - 5);
```
Comment (lynnmunday) [inline modules/optimization/test/src/optimizationreporters/QuadraticMinimizeConstrained.C:74]: Is this constraint a hard coded 5?  Can it go in the input file or some way to make this more clear.

File: modules/optimization/test/tests/executioners/constrained/equality/quadratic_minimize_constrained.i [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+[]
+
+[OptimizationReporter]
+  type = QuadraticMinimizeConstrained
+  parameter_names = 'results'
+  num_values = 3
+  initial_condition = '1 1 1'
+  objective = 0.0
+  solution = '1 2 3'
+  measurement_points = ''
```
Comment (lynnmunday) [inline modules/optimization/test/tests/executioners/constrained/equality/quadratic_minimize_constrained.i:11]: A parameter here for the solution_sum_equality constraint setting it to 5 will make it clear why the gold values add up to 5 instead of 6.

File: modules/optimization/test/tests/executioners/constrained/inequality/forward_and_adjoint.i [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+[VectorPostprocessors]
+  [grad_bc_left]
+    type = SideOptimizationNeumannFunctionInnerProduct
+    variable = temperature_adjoint
+    function = left_function
+    boundary = left
+    execute_on = ADJOINT_TIMESTEP_END
+  []
+  [gradient_c]
+    type = VectorOfPostprocessors
```
Comment (lynnmunday) [inline modules/optimization/test/tests/executioners/constrained/inequality/forward_and_adjoint.i:107]: I think it would be helpful to move all of the inequality constraint set-up to its own block at the bottom of the input file so that I can tell what is different about this compared to a regular input file.  It would be even better if you could use the multiple input file stuff and just have a small input file containing your additions but I don't know if something like this would work.  Also model.i should be a link back to the unconstrained input file. 
```
[MultiApps]
  [forward]
    type = FullSolveMultiApp
    input_files = model.i model_InEqConstraints.i
    execute_on = FORWARD
  []
```

Or something like this to seperate the constraints into their own blocks.
```suggestion
#---------Inequality constraints------------#
[VectorPostprocessors]
  [gradient_c]
    type = VectorOfPostprocessors
    postprocessors = 'dc_da dc_db'
  []
  [ineq]
    type = VectorOfPostprocessors
    postprocessors = 'constraint'
    execute_on = TIMESTEP_END
  []
[]
[Postprocessors]
  [sum]
    type = FunctionSideIntegral
    boundary = left
    function = left_function

  []
  [constraint]
    type = ParsedPostprocessor
    function = '150 - sum' # 150 is the constraint we want to satisfy
    pp_names = sum
    execute_on = TIMESTEP_END
  []
  [dc_da]
    type = FunctionSideIntegral
    boundary = left
    function = -1
  []
  [dc_db]
    type = FunctionSideIntegral
    boundary = left
    function = '-y'
  []
[]  
```

File: modules/optimization/test/tests/executioners/constrained/inequality/tests [hunk 7]
```diff
@@ -0,0 +1,11 @@
+[Tests]
+  design = 'Optimize.md'
+  [inequality_constraint]
+    issues = '#24930'
+    max_threads = 1 # Optimize executioner does not support multiple threads
+    input = main_auto_adjoint.i
+    jsondiff = main_auto_adjoint_out_forward0.json
+    type = JSONDiff
+    requirement = "The system shall minimize a quadratic objective function from a test object."
```
Comment (lynnmunday) [inline modules/optimization/test/tests/executioners/constrained/inequality/tests:9]: I think you also need:

```suggestion
# steady solve
   recover = false
    requirement = "The system shall minimize a quadratic objective function from a test object."
```

File: modules/optimization/test/tests/executioners/constrained/shape_optimization/forward.i [hunk 8]
```diff
@@ -0,0 +1,278 @@
+inner_radius = 6
```
Comment (lynnmunday) [inline modules/optimization/test/tests/executioners/constrained/shape_optimization/forward.i:1]: Add a comment to the top of this file that it is documented as an example and should not be changed without updating the documentation.

File: modules/optimization/test/tests/executioners/constrained/shape_optimization/tests [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+  design = 'GeneralOptimization.md'
+  [inequality_constraint]
+    issues = '#24930'
+    max_threads = 1 # Optimize executioner does not support multiple threads
+    input = main.i
+    exodiff = main_out_forward0.e
+    type = Exodiff
+    requirement = "The system shall be able to minimize a constrained shape optimization problem using the ALMM algorithm."
+    heavy=true
+    allow_warnings=true
```
Comment (lynnmunday) [inline modules/optimization/test/tests/executioners/constrained/shape_optimization/tests:11]: Maybe add 
```
   recover = false
```
I don't know if that is needed for heavy tests or not.

File: modules/optimization/test/tests/executioners/constrained/inequality/measurementData.csv [hunk 10]
```diff
@@ -0,0 +1,3 @@
+measured_value,coordx,y,z
+207,0.2,0.2,0
```
Comment (lynnmunday) [inline modules/optimization/test/tests/executioners/constrained/inequality/measurementData.csv:2]: Don't you need all four measurement points?  I'm surprised this works, isn't it underconstrained.  You can skip the csv file and put the measurements directly into the optimizationReporter like I showed above.

File: modules/optimization/test/tests/executioners/constrained/inequality/main_auto_adjoint.i [hunk 11]
```diff
@@ -0,0 +1,77 @@
+# This tests that a linear and constant function can be scaled in
```
Comment (lynnmunday) [inline modules/optimization/test/tests/executioners/constrained/inequality/main_auto_adjoint.i:1]: This comment to should be changed to something like this and also include something about the constraint.

```suggestion
# This tests that a linear function can be scaled in
# functionNeumannBCs applied to the left sideset using
# a parsed functions.  The scale of the linear function
# are being parameterized.
```

File: modules/optimization/test/tests/executioners/constrained/inequality/main_auto_adjoint.i [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+  type = OptimizationReporter
+  parameter_names = 'left'
+  num_values = '2'
+  initial_condition = '10 10'
+  lower_bounds = '0'
+  upper_bounds = '1000'
+
+  inequality_names = 'ineq'
+
+  measurement_file = 'measurementData.csv'
```
Comment (lynnmunday) [inline modules/optimization/test/tests/executioners/constrained/inequality/main_auto_adjoint.i:19]: I would put the measurements into the file.  I also would include all four measurements, I think a single measurement would be underconstrained.  
```suggestion
  measurement_points = '0.2 0.2 0
                        0.8 0.6 0
                        0.2 1.4 0
                        0.8 1.8 0'
  measurement_values = '207 204 185 125'
```
You'll need to update these with synthetic measurements for your problem.

**Review by lynnmunday — CHANGES_REQUESTED** [review]

File: modules/optimization/test/tests/executioners/constrained/shape_optimization/tests [hunk 13]
```diff
@@ -4,10 +4,11 @@
     issues = '#24930'
     max_threads = 1 # Optimize executioner does not support multiple threads
     input = main.i
-    exodiff = main_out_forward0.e
-    type = Exodiff
+    csvdiff = main_out.csv
```
Comment (lynnmunday) [inline modules/optimization/test/tests/executioners/constrained/shape_optimization/tests:7]: I think you'll need tolerances on this.

File: framework/src/parser/Parser.C [hunk 14]
```diff
@@ -1524,6 +1524,7 @@ Parser::extractParams(const std::string & prefix, InputParameters & p)
         setvectorvector(DistributionName);
         setvectorvector(SamplerName);
         setvectorvector(TagName);
+        setvectorvector(ReporterValueName);
```
Comment (maxnezdyur) [inline framework/src/parser/Parser.C:1527]: Needed to add this to have a vector of vector of reporter names.

File: modules/optimization/test/include/optimizationreporters/QuadraticMinimizeConstrained.h [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+#pragma once
+
+#include "OptimizationReporter.h"
+
+/**
+ * This form function simply represents a quadratic objective function:
+ *    f(x) = val + \sum_{i=1}^N (x_i - a_i)^2
+ * where val is the input objective value and a_i is the input solution.
+ */
+class QuadraticMinimizeConstrained : public OptimizationReporter
```
Comment (zachmprince) [inline modules/optimization/test/include/optimizationreporters/QuadraticMinimizeConstrained.h:19]: Why can't this inherit from `QuadraticMinimize`?

File: modules/optimization/test/src/optimizationreporters/QuadraticMinimizeConstrained.C [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+{
+
+  unsigned int i = 0;
+  for (const auto & param : _parameters)
+  {
+    Real equality_constraint = 0;
+    for (const auto & val : *param)
+    {
+      equality_constraint += val;
+    }
```
Comment (zachmprince) [inline modules/optimization/test/src/optimizationreporters/QuadraticMinimizeConstrained.C:76]:
```suggestion
    const Real equality_constraint = std::accumulate(param->begin(), param->end(), 0.0);
```

File: modules/optimization/test/src/optimizationreporters/QuadraticMinimizeConstrained.C [hunk 17]
```diff
@@ -0,0 +1,91 @@
...
+  }
+  eqs_constraints.close();
+}
+void
+QuadraticMinimizeConstrained::computeEqualityGradient(libMesh::PetscMatrix<Number> & gradient) const
+{
+  gradient.zero();
+
+  for (unsigned long i = 0; i < _n_eq_cons; i++)
+    for (unsigned long j = 0; j < _parameters[0]->size(); j++)
```
Comment (zachmprince) [inline modules/optimization/test/src/optimizationreporters/QuadraticMinimizeConstrained.C:87]:
```suggestion
  for (const auto i : make_range(_n_eq_cons))
    for (const auto j : index_range(_parameters[0]))
```

**Review by zachmprince — CHANGES_REQUESTED** [review]
This my initial review of the code. I'll go through the documentation and tests later.

File: modules/optimization/src/optimizationreporters/GeneralOptimization.C [hunk 18]
```diff
@@ -0,0 +1,114 @@
...
+{
+  InputParameters params = OptimizationReporterBase::validParams();
+
+  params.addRequiredParam<ReporterValueName>("objective_name", "Objective reporter name.");
+  params.addParam<std::vector<dof_id_type>>(
+      "num_values",
+      "Number of parameter values associated with each parameter group in 'parameter_names'.");
+  params.addParam<ReporterValueName>("num_values_name",
+                                     "Reporter that holds the number of parameter values "
+                                     "associated with each parameter group in 'parameter_names'.");
```
[the remark this answers is not in the document]
Comment (maxnezdyur) [reply modules/optimization/src/optimizationreporters/GeneralOptimization.C:29]: I am thinking of two scenarios. 
1: The number of parameters is set regardless of any
changes in the forward problem aka mesh refinement does not affect number of
parameters.
2: The number of parameters changes as the mesh gets refined for example
individual elemental material properties. 
When I say mesh gets refined, I don't mean in the middle of the solve, just a
totally new run. Having option 2 makes it so the user doesn't have to run the
forward problem get the number of parameters of the new mesh then update that in
the optimization app.

File: modules/optimization/src/optimizationreporters/OptimizationReporterBase.C [hunk 19]
```diff
@@ -20,6 +21,10 @@ OptimizationReporterBase::validParams()
       "parameter_names", "List of parameter names, one for each group of parameters.");
   params.addRangeCheckedParam<Real>(
       "tikhonov_coeff", 0.0, "tikhonov_coeff >= 0", "Coefficient for Tikhonov Regularization.");
+  params.addParam<std::vector<ReporterValueName>>(
+      "equality_names", std::vector<ReporterValueName>(), "List of equality names.");
+  params.addParam<std::vector<ReporterValueName>>(
+      "inequality_names", std::vector<ReporterValueName>(), "List of inequality names.");
```
[the remark this answers is not in the document]
Comment (maxnezdyur) [reply modules/optimization/src/optimizationreporters/OptimizationReporterBase.C:27]: Possible use cases:
Find the correct force on the boundary but constraining the sum on the boundary
to less than a known value.
Using double-well potential constraint where you know the parameters can only be
one of two values.
Whether it is relevant now to the current inverse optimization prolems, I can't say.

[the remark this answers is not in the document]
Comment (lynnmunday) [reply modules/optimization/src/optimizationreporters/OptimizationReporterBase.C:27]: This inverse optimization test uses it:   modules/optimization/test/tests/executioners/constrained/inequality/forward_and_adjoint.i

[the remark this answers is not in the document]
Comment (zachmprince) [reply modules/optimization/src/optimizationreporters/OptimizationReporterBase.C:27]: Fair enough!

File: modules/optimization/src/optimizationreporters/GeneralOptimization.C (hunk 18 above)
[the remark this answers is not in the document]
Comment (zachmprince) [reply modules/optimization/src/optimizationreporters/GeneralOptimization.C:29]: Would there be a useful error message if the number of parameters didn't match what is expected in the forward problem?

[the remark this answers is not in the document]
Comment (maxnezdyur) [reply modules/optimization/src/optimizationreporters/GeneralOptimization.C:29]: Most likely the forward problem would have to output that message. I don't think
I am 100% understanding what you are asking, though. For scenario 2, the forward
problem determines the number of parameters so there shouldn't be a mismatch.
For scenario 1, it would have the same error message that would happen normally
if specifying incorrect number of parameters.

[the remark this answers is not in the document]
Comment (zachmprince) [reply modules/optimization/src/optimizationreporters/GeneralOptimization.C:30]: I guess I'm fine with this. I hope in the future this can be setup automatically.

File: modules/optimization/doc/content/modules/optimization/examples/shapeOpt_Annulus.md [hunk 20]
```diff
@@ -0,0 +1,122 @@
+# Constrained Optimization Example: Shape Optimization
+
+## Background
```
Comment (zachmprince) [inline modules/optimization/doc/content/modules/optimization/examples/shapeOpt_Annulus.md:3]: I would suggest having a problem description section before jumping into the example. Explain the geometry, the physics, and what is trying to be optimized.

File: modules/optimization/doc/content/modules/optimization/examples/shapeOpt_Annulus.md [hunk 21]
```diff
@@ -0,0 +1,122 @@
...
+
+The MOOSE Optimization module provides a flexible framework for solving
+optimization problems in MOOSE. This page will demonstrate constrained shape
+optimization of an annulus while satisfying an equality volume constraint.
```
Comment (zachmprince) [inline modules/optimization/doc/content/modules/optimization/examples/shapeOpt_Annulus.md:7]: It would be great if you could add a few sentences on what shape optimization is in general terms. Adding a citation would help a lot too.

File: modules/optimization/doc/content/modules/optimization/examples/shapeOpt_Annulus.md [hunk 22]
```diff
@@ -0,0 +1,122 @@
...
+
+# Example: Shape Optimization
```
Comment (zachmprince) [inline modules/optimization/doc/content/modules/optimization/examples/shapeOpt_Annulus.md:9]: I would reorganize this section a little bit. Under the main header, state the algorithm and multiapp structure. Then I would go with the forward problem description first, then the optimization app.

File: modules/optimization/doc/content/modules/optimization/examples/shapeOpt_Annulus.md [hunk 23]
```diff
@@ -0,0 +1,122 @@
...
+
+## Main-App Optimization
+
+Optimization problems are solved using the [MultiApps](MultiApps/index.md)
+system.  The main application contains the optimization executioner and the
+sub-applications solve the forward and adjoint PDE.   The main application input
```
Comment (zachmprince) [inline modules/optimization/doc/content/modules/optimization/examples/shapeOpt_Annulus.md:15]: It doesn't contain the adjoint, right?

File: modules/optimization/doc/content/modules/optimization/examples/shapeOpt_Annulus.md [hunk 24]
```diff
@@ -0,0 +1,122 @@
...
+is shown in [main_app].
+
+!listing test/tests/executioners/constrained/shape_optimization/main.i
+         id=main_app
+         caption= Constrained Optimization Main App
```
Comment (zachmprince) [inline modules/optimization/doc/content/modules/optimization/examples/shapeOpt_Annulus.md:20]: I don't think this listing is necessary.

File: modules/optimization/doc/content/modules/optimization/examples/shapeOpt_Annulus.md [hunk 25]
```diff
@@ -0,0 +1,122 @@
...
+[!param](/Executioner/Optimize/petsc_options_value).
+constraint support. This example computes the objective function's gradient using finite differencing in TAO with the `petsc_options_iname` `-tao_fd`.  Finite differencing only uses the forward problem's objective value and does not require an adjoint solve.
+`-tao_almm_subsolver_tao_type` and  `bqnktr` set the subsolver to the Bounded Quasi-Newton-Krylov
+Trust Region Method.
+
+!listing test/tests/executioners/constrained/shape_optimization/main.i
+        block= Executioner
+        id=exec
+        caption= Main application `Executioner` block
+
```
Comment (zachmprince) [inline modules/optimization/doc/content/modules/optimization/examples/shapeOpt_Annulus.md:54]: Could you add a paragraph and listing describing the `MultiApps` and `Transfers` blocks.

File: modules/optimization/doc/content/modules/optimization/examples/shapeOpt_Annulus.md [hunk 26]
```diff
@@ -0,0 +1,122 @@
...
+         constraints
+
+To enable constrained optimization the
+[!param](/Executioner/Optimize/tao_solver) needs to be set to `taoalmm`, which
+is an Augmented Lagrangian method. This is the only TAO solver in MOOSE that has
+constraint support. The `taoalmm` subsolver is set with [!param](/Executioner/Optimize/petsc_options_iname) and
+[!param](/Executioner/Optimize/petsc_options_value).
+constraint support. This example computes the objective function's gradient using finite differencing in TAO with the `petsc_options_iname` `-tao_fd`.  Finite differencing only uses the forward problem's objective value and does not require an adjoint solve.
+`-tao_almm_subsolver_tao_type` and  `bqnktr` set the subsolver to the Bounded Quasi-Newton-Krylov
+Trust Region Method.
```
Comment (zachmprince) [inline modules/optimization/doc/content/modules/optimization/examples/shapeOpt_Annulus.md:48]: It would be good to mention what all the other parameters mean, or at least link to the petsc website describing them.

File: modules/optimization/doc/content/modules/optimization/examples/shapeOpt_Annulus.md [hunk 27]
```diff
@@ -0,0 +1,122 @@
...
+
+In the forward app, a heat conduction problem is solved with convective flux
+type boundary condition that changes with the inner radius. The outer radius is
+insulated and there is a constant source term throughout. The objective is
+find the two radii that will have the lowest max temperature while satisfying a
+volume constraint.
+
+!listing test/tests/executioners/constrained/shape_optimization/forward.i
+         id=forward_app
+         caption= Constrained Optimization Forward SubApp
```
Comment (zachmprince) [inline modules/optimization/doc/content/modules/optimization/examples/shapeOpt_Annulus.md:65]: I really wouldn't put the entire input file in one listing. It's better to go through each feature one-by-one. One can always click the link in the listing to view the entire file.

File: modules/optimization/doc/content/modules/optimization/examples/shapeOpt_Annulus.md [hunk 28]
```diff
@@ -0,0 +1,122 @@
...
+
+In the forward app, the two optimization parameters control the inner and outer
+values are used to displace the boundaries of the
+[ParsedOptimizationFunction](/Functions/ParsedOptimizationFunction) to calculate
+the cartesian boundary displacement. These
+values are used displace the boundaries of the
+annulus to the correct position using Dirichlet boundary conditions. A diffusion
+problem on the displacements then allow for the interior elements to be
+smoothed. For the temperature field, the physics is solved on the displaced mesh
+to calculate the current objective.
```
Comment (zachmprince) [inline modules/optimization/doc/content/modules/optimization/examples/shapeOpt_Annulus.md:75]: Smoothing the displacements using a diffusion problem is quite interesting. Is there any precedent for this? I suspect it makes the forward problem a little more difficult to solve.

File: modules/optimization/doc/content/modules/optimization/examples/shapeOpt_Annulus.md [hunk 29]
```diff
@@ -0,0 +1,122 @@
...
+
+For the equality constraints, a gradient of the constraint with respect to the
+parameters is needed for TAO. The gradient is analytically computed for the
+current radii.
+
+!listing test/tests/executioners/constrained/shape_optimization/forward.i
+         block= Functions BCs Kernels
+         id=forward_app_shape
+         caption= Forward App Shape Optimization
```
Comment (zachmprince) [inline modules/optimization/doc/content/modules/optimization/examples/shapeOpt_Annulus.md:84]: I would put this above the previous paragraph. It doesn't seem to be related to constraints.

File: modules/optimization/doc/content/modules/optimization/examples/shapeOpt_Annulus.md [hunk 30]
```diff
@@ -0,0 +1,122 @@
...
+
+!listing test/tests/executioners/constrained/shape_optimization/forward.i
+         block=Postprocessors
+         id=forward_app_obj
+         caption= Forward App Objective Calculations
+
+## Adjoint Problem Sub-App
+
+Since we are using finite-differencing to determine the gradients of our
+objective function, a Adjoint Sub-App is not needed.
```
Comment (zachmprince) [inline modules/optimization/doc/content/modules/optimization/examples/shapeOpt_Annulus.md:101]: This is disappointing. Any reason we can't use the adjoint to compute the gradient?

File: modules/optimization/doc/content/modules/optimization/examples/shapeOpt_Annulus.md [hunk 31]
```diff
@@ -0,0 +1,122 @@
...
+
+After optimizing the design to decrease the max
+temperature, the new design has a max temperature around 160 and the volume
+constraint was satisfied to the tolerance. [opt_sol] shows the optimized designs
+and the new temperature profile.
+
+!media large_media/optimization/constrainedOpt_shapeSoln.png
+        style=width:60%;margin:auto;padding-top:2.5%;background-color: white;color: black;
+        id=opt_sol
+        caption= Optimized Annulus temperature profile
```
Comment (zachmprince) [inline modules/optimization/doc/content/modules/optimization/examples/shapeOpt_Annulus.md:121]: I would put some rulers on the figure to show what the resulting outer radius and thickness of the annulus is. Or you can describe in the text. Or maybe even put both figures on the same plot so they are scaled the same.

**Review by zachmprince — DISMISSED** [review]
Amazing work Max!

File: modules/optimization/doc/content/modules/optimization/examples/shapeOpt_Annulus.md (hunk 28 above)
Comment (maxnezdyur) [reply modules/optimization/doc/content/modules/optimization/examples/shapeOpt_Annulus.md:75]: It would at adds one more nonlinear iteration, usually. The diffusion problem is linear so it solves after the first NL iteration then the rest of the physics solves at the "same" rate. IIRC I think malamute does this for their ALE mesh movement.

File: modules/optimization/doc/content/modules/optimization/examples/shapeOpt_Annulus.md (hunk 30 above)
Comment (maxnezdyur) [reply modules/optimization/doc/content/modules/optimization/examples/shapeOpt_Annulus.md:101]: The objective is the maximum nodal temperature, which would not have a continuous gradient. Also one point of this example is to show that nearly any MOOSE app can be converted to a optimization problem using FD and a post processor.

File: modules/optimization/doc/content/modules/optimization/examples/shapeOpt_Annulus.md (hunk 27 above)
Comment (maxnezdyur) [reply modules/optimization/doc/content/modules/optimization/examples/shapeOpt_Annulus.md]: I got rid of the two entire input file listing and fixed up some of the wording!

File: modules/optimization/doc/content/modules/optimization/examples/shapeOpt_Annulus.md (hunk 31 above)
Comment (maxnezdyur) [reply modules/optimization/doc/content/modules/optimization/examples/shapeOpt_Annulus.md:166]: I described it in text, didn't realize that GitHub didn't mark is as outdated.

File: modules/optimization/doc/content/modules/optimization/examples/shapeOpt_Annulus.md (hunk 29 above)
Comment (maxnezdyur) [reply modules/optimization/doc/content/modules/optimization/examples/shapeOpt_Annulus.md:83]: Flipped the paragraphs around.

**Comment by oanaoana** [comment]
It seems I fixed the problem by just adjusting and moving things around to be properly parsed. The problem was MOOSE, not TAO, but something is error prone on the reading options side.  After running valgrind with gdb and tons of options like `monitor leak_check` etc it trickles down to libmesh, specifically boost is triggered. This is the MOOSE exit point `ConsoleStream.C:18`.  Let me know if to PR to your branch or MOOSE directly (I checked out your branch and rebased so not sure if there was more work in it)

**Comment by maxnezdyur** [comment]
I don't care if you PR to my branch or MOOSE directly. I guess it would be best to PR to MOOSE directly to not mix PRs. I guess this would be a general fix that should be separated(not sure what the general consensus is). Thanks for finding this fix!

**Comment by maxnezdyur** [comment]
@lynnmunday 
The test failures don't seem to have anything to do with the PR, but rerunning them doesn't seem to help.

File: modules/optimization/test/tests/optimizationreporter/general_opt/main.i [hunk 32]
```diff
@@ -0,0 +1,58 @@
...
+[]
+
+[Postprocessors]
+  [cost_pp]
+    type = Receiver
+  []
+[]
+[Problem]
+  solve = false
+[]
```
Comment (lynnmunday) [inline modules/optimization/test/tests/optimizationreporter/general_opt/main.i:18]: what are these for?

File: modules/optimization/test/tests/optimizationreporter/general_opt/tests [hunk 33]
```diff
@@ -0,0 +1,14 @@
+[Tests]
```
Comment (lynnmunday) [inline modules/optimization/test/tests/optimizationreporter/general_opt/tests:1]: This is a good test, It took me a while to figure out what you were doing.  Should we just make the quadraticMinimize test like this?

**Review by lynnmunday — APPROVED** [review]
Don't make any code changes based on my comments, I just thought of them as refactors we can talk about doing at some point.

File: modules/optimization/include/optimizationreporters/OptimizationReporterBase.h [hunk 34]
```diff
@@ -94,6 +133,24 @@ class OptimizationReporterBase : public GeneralReporter
   /// Tikhonov Coefficient for regularization
   const Real _tikhonov_coeff;
 
+  /// Equality constraint names
+  const std::vector<ReporterValueName> * _equality_names;
```
[the remark this answers is not in the document]
Comment (maxnezdyur) [reply modules/optimization/include/optimizationreporters/OptimizationReporterBase.h:137]: Its a pointer so that we only create data structures when using the ALMM approach.

File: modules/optimization/include/optimizationreporters/OptimizationReporterBase.h [hunk 35]
```diff
@@ -81,6 +117,9 @@ class OptimizationReporterBase : public GeneralReporter
    */
   std::vector<Real> fillParamsVector(std::string type, Real default_value) const;
 
+  /// Sets the initial conditions and bounds right before it is needed.
+  virtual void setICsandBounds(){};
```
[the remark this answers is not in the document]
Comment (maxnezdyur) [reply modules/optimization/include/optimizationreporters/OptimizationReporterBase.h:121]: No we need to call this function in the base class `setInitialCondition` method. We allow any child class to override because they have different ICs and may need to set them differently.

File: modules/optimization/test/tests/optimizationreporter/general_opt/main.i (hunk 32 above)
Comment (maxnezdyur) [reply modules/optimization/test/tests/optimizationreporter/general_opt/main.i]: Got rid the post processor nice call.

File: modules/optimization/test/tests/optimizationreporter/general_opt/tests (hunk 33 above)
Comment (maxnezdyur) [reply modules/optimization/test/tests/optimizationreporter/general_opt/tests:1]: I think this is cleaner way because it allows to try different optimization options with recompiling or creating multiple test classes. It also is a good test to show how "general" GeneralOptimization is.

**Review by lynnmunday — APPROVED** [review]
Great!

**Comment by lynnmunday** [comment]
@recuero Will you review and merge this?  @zachmprince and I both reviewed it.  Zach would have merged it before he left but there was a problem with a PETSC version update that occurred during this PR.  This has been fixed by the git sha [11876fb](https://github.com/idaholab/moose/pull/24934/commits/11876fbb30540dd5cc4372a24391dd425bcfc40e)

**Comment by recuero** [comment]
Will do.

File: modules/optimization/examples/shape_optimization/forward.i [hunk 36]
```diff
@@ -0,0 +1,278 @@
...
+  [diff_coef]
+    type = ParsedMaterial
+    coupled_variables = 'dist_between'
+    expression = '1/(dist_between+0.5)'
+    property_name = 'diff_coef'
+  []
+  [h]
+    type = ADGenericFunctionMaterial
+    prop_names = h
+    prop_values = h
```
Comment (recuero) [inline modules/optimization/examples/shape_optimization/forward.i:109]: Can you add syntax checking testing to the examples? (I take it that regular regression testing may not be possible since they may take longer).

Comment (recuero) [reply modules/optimization/examples/shape_optimization/forward.i:109]: Wait... You do have a gold file there, but I can't see `tests`.

File: modules/optimization/include/optimizationreporters/GeneralOptimization.h [hunk 37]
```diff
@@ -0,0 +1,37 @@
...
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "OptimizationReporterBase.h"
+#include "libmesh/id_types.h"
+#include "libmesh/libmesh_common.h"
+
+/**
+ */
```
Comment (recuero) [inline modules/optimization/include/optimizationreporters/GeneralOptimization.h:17]: Add some description here.

File: modules/optimization/examples/shape_optimization/forward.i (hunk 36 above)
Comment (maxnezdyur) [reply modules/optimization/examples/shape_optimization/forward.i:109]: Nice catch. I have a test for shape optimization that I also use for the shape opt example. This example is no longer needed and I got rid of it.

**Review by recuero — APPROVED** [review]

**Comment by lynnmunday** [comment]
@loganharbour will you merge this since it is approved?

**Comment by recuero** [comment]
I'll merge tomorrow.

**Comment by recuero** [comment]
I am going to dismiss @zachmprince review, it seems his review has been addressed.
